code/game/objects/items/weapons/rpd.dm

```diff
@@ -0,0 +1,292 @@
+	else if(href_list["pipetype"])
+		pipetype = text2num(sanitize(href_list["pipetype"]))
+	else if(href_list["mode"])
+		mode = text2num(sanitize(href_list["mode"]))
+	else
+		return
+	nanomanager.update_uis(src)
+
+//What the RPD actually does
+
+/obj/item/weapon/rpd/afterattack(atom/target, mob/user, proximity)
+	..()
+	var/turf/T = get_turf(target)
+	if(ismob(target) || istype(target, /obj/structure/window) || !proximity || cooldown)
+		return
+	if(mode == ATMOS_MODE)
+		if(istype(T, /turf/simulated/wall)) //Drilling into walls takes time and we don't want to do anything else while drilling
+			cooldown = TRUE
+			playsound(get_turf(src), "sound/weapons/circsawhit.ogg", 50, 1)
+			user.visible_message("<span class = 'notice'>[user] starts drilling a hole in [T]...</span>", "<span class = 'notice'>You start drilling a hole in [T]...</span>", "<span class = 'warning'>You hear a drill.</span>")
+			if(!do_after(user, walldelay, target = T))
+				cooldown = null
+				return
+			user.visible_message("<span class = 'notice'>[user] finishes drilling a hole in [T]!</span>", "<span class = 'notice'>You finish drilling a hole in [T]!</span>", "<span class = 'warning'>You hear clanking.</span>")
+		if(whatpipe == GAS_SENSOR)//need to check if it's a meter or a gas sensor separately ayylmao
+			new /obj/item/pipe_gsensor(T)
+			to_chat(user, "<span class = 'notice'>[src] dispenses a sensor!</span>")
+		else if(whatpipe == METER)
+			new /obj/item/pipe_meter(T)
+			to_chat(user, "<span class = 'notice'>[src] dispenses a meter!</span>")
+		else
+			var/obj/item/pipe/P = new(T, pipe_type = whatpipe, dir = user.dir) //Now we make the pipes
+			if(iconrotation == 0 && Checkifbent(P.pipe_type)) //Automatic rotation of dispensed pipes
+				P.dir = turn(user.dir, 135)
+			else if(iconrotation == 0 && P.pipe_type in list(CONNECTOR, UNARY_VENT, GAS_SCRUBBER, HE_EXCHANGER, SIMPLE_CAP, SUPPLY_CAP, SCRUBBERS_CAP, INJECTOR, PASSIVE_VENT)) //Some pipes dispense oppositely to what you'd expect, but we don't want to do anything if they selected a direction
+				call(P, /obj/item/pipe/verb/flip)()
+			else if(iconrotation != 0 && Checkifbent(P.pipe_type)) //If they selected a rotation and the pipe is bent
+				P.dir = turn(iconrotation, -45)
+			else if(iconrotation != 0)
+				P.dir = iconrotation
+			P.update()
+			to_chat(user, "<span class = 'notice'>[src] rapidly dispenses [P]!</span>")
+		Activaterpd(1)
+	else if(mode == DISPOSALS_MODE)
+		if(istype(T, /turf/simulated/wall))
+			to_chat(user, "<span class = 'warning'>That type of pipe won't fit on [T]!</span>")
+			return
+		if(is_blocked_turf(T) && Checkifdense(whatdpipe))
+			to_chat(user, "<span class = 'warning'>You can't dispense that on [T] because something is in the way!</span>")
+			return
+		var/obj/structure/disposalconstruct/P = new(T) //Now we make the pipe
+		P.dir = iconrotation
+		P.ptype = whatdpipe
+		if(Checkifdense(whatdpipe)) //To properly set density of our bins/chutes/outlets
+			P.density = 1
```

Comment:
I might remove this so that blocking off areas with masses of bins is not an issue. Or maybe I won't 👀